scripts/gcloud-sql-export.sh

```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+set -o nounset -o errexit -o pipefail
+[[ -n "${TRACE:-}" ]] && set -x
+
+function gcloud_activate_service_account() {
+  local file="$1"
+
+  gcloud auth activate-service-account --key-file="$file"
+  gcloud config set project "$(jq -r .project_id "$file")"
+}
+
+
+DATESTAMP=$(date -Iseconds) 
+GCS_BACKUP_PATH="gs://${GOOGLE_SQL_BACKUP_BUCKET}/${GOOGLE_SQL_BACKUP_BUCKET_PATH}/${DATESTAMP}_${GOOGLE_SQL_INSTANCE_NAME}_${DATABASE}.gz"
+
+gcloud_activate_service_account "${GOOGLE_APPLICATION_CREDENTIALS}"
+
+gcloud --verbosity=debug sql export sql "${GOOGLE_SQL_INSTANCE_NAME}" "${GCS_BACKUP_PATH}" --database "${DATABASE}"
```

Comment:
have we checked how does this command behave when there is an already running backup ? 

Also, are we going to check in the script that a "successful" backup does not exist yet with a date that is less then the current schedule for this job ?

Reply:
Yes for the what happens when both are run - I tracked this on our local gitlab case.

`gcloud` operations cannot run concurrently and a 403 error is returned. That causes an error-exit status, so K8s does the right thing.

I can still do the successful backup check - maybe make it an optional flag.

Reply:
@primeroz I think any check would need to be fairly limited in scope.

There are 2 problems here:

- We would need to translate the cronjob schedule setting, which is going to be tricky.
- There's no server-side filtering on metadata
- Doing such checks puts constraints on the directory layout of our objects (yyyy/mm/dd etc)

I would say anyhting we do in this area needs to be very limited in scope.

Reply:
> We would need to translate the cronjob schedule setting, which is going to be tricky. 

If we use the string `@hourly` or `@daily` it would be easier since you would not need conversion https://kubernetes.io/docs/tasks/job/automated-tasks-with-cron-jobs/#schedule 

in that case though it would only possible to run jobs at "0 * * * *" and "0 0 * * *"

I see the issue , the problem is that if we don't do any check we risk issues ... i guess that is the difference between a command `mysqldump or gcloud sql export` and `restic` which is instead a backup system.

Not sure what to say, if we do not care then let's not care 

> gcloud operations cannot run concurrently and a 403 error is returned. That causes an error-exit status, so K8s does the right thing.

does the job fail when we encounter an error from gcloud ? 
* If it does then when we get 403 k8s will retry
* if it does not then we risk loosing backups because of transient gcloud errors

Reply:
It does fail yeah. 

I also see the issue and think we need todo something.

I would suggest we support `@daily` for now since that's required for portal (maybe hourly too).

In addition, this means that the filenames should be fairly strict (and not down minutes/seconds).

That way we can look for explicit filenames that should exist (and avoid any filtering).

Also, we might want to consider a `yyyy/mm/dd/backup.gz` format, if we do require filtering, it will help somewhat.

Reply:
i see we are using `2020-03-13T12:36:14+0000_` as prefix ... is it not easier to use a Unix Epoch ? :)

Reply:
I don't think it's easier - it's not very human readable, and the current script outputs INFO on backups that it tries to create (or skip) - this could be useful for debugging issues with backups.

If we consider a DR scenario, then having a readable timestamp will make it easier to find the backup that we need to restore from without having to convert from a unix-timestamp.